### PR TITLE
os/bluestore: make pinning logic simpler

### PR DESCRIPTION
**Pull request by aclamk**
This is a follow-up of  #44311, sitting on top of it.

Branch with more ceph_assert wip-aclamk-cleanup-onode-pin-test is going to teuthology for testing.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -3642,12 +3635,8 @@ void BlueStore::Onode::get() {
       ocs->lock.lock();
     }
     bool was_pinned = pinned;
-    pinned = nref >= 2;
-    // additional increment for newly pinned instance
+    pinned = nref >= 1;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: Looks like I'm wrong at this line - it still should be nref >= 2
Going to re-test and update the original PR.

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -3623,16 +3623,9 @@ void BlueStore::Onode::calc_omap_tail(
...
-// increment when newly pinned instance is detected. And -1 on unpin.
-// This prevents from a conflict with a delete call (when nref == 0).
-// The latter might happen while the thread is in unpin() function
-// (and e.g. waiting for lock acquisition) since nref is already
-// decremented. And another 'putting' thread on the instance will release it.
-//
+
 void BlueStore::Onode::get() {
-  if (++nref >= 2 && !pinned) {
+  if (++nref >= 1 && !pinned) {
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc]: And this should stay >=2 too.

**Comment by ifed01** [comment]
@aclamk - I fixed the condition for 'pinned' state - this should be nref >= 2 not nref > =1 as I proposed initially.

**Comment by markhpc** [comment]
Adding performance flag just because it touches fairly performance sensitive code (even if we don't expect it to have an impact)

**Comment by aclamk** [comment]
@ifed01 You mentioned this as preferred solution for onode pinning cleanup.
I am not fully clear at this point - does this PR need any more attention?

**Comment by ifed01** [comment]
> @ifed01 You mentioned this as preferred solution for onode pinning cleanup. I am not fully clear at this point - does this PR need any more attention?

yes, it needs rebase (and probably some rework) since my original PR has been updated and merged.

File: src/os/bluestore/BlueStore.cc (hunk 2 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: no need any more indeed

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -3686,21 +3687,24 @@ void BlueStore::Onode::put() {
...
       ocs->lock.lock();
     }
-    bool need_unpin = pinned;
-    pinned = pinned && nref >= 2;
-    need_unpin = need_unpin && !pinned;
-    if (cached && need_unpin) {
+    // on 1st glance pinned == true always
+    // on 2nd inspection we see that some other thread could have beaten us to unpinning
+    bool was_pinned = pinned;
+    if (nref >= 2) ceph_assert(pinned);
```
[the remark this answers is not in the document]
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc]: not a big deal since this goes away in the last commit.

**Review by ifed01 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**